### PR TITLE
Give specific instruction for zsh on macOS for global tools PATH

### PR DESCRIPTION
**Pull request by wli3**
fix https://github.com/dotnet/cli/issues/9321
example:

```
% dotnet tool install -g dotnetsay --add-source https://api.nuget.org/v3/index.json
Tools directory '/Users/williamli/.dotnet/tools' is not currently on the PATH environment variable.
If you are using zsh, you can add it to your profile by running the following command:

cat << \EOF >> ~/.zprofile
# Add .NET Core SDK tools
export PATH="$PATH:/Users/williamli/.dotnet/tools"
EOF

And run `zsh -l` to make it available for current session.

You can only add it to the current session by running the following command:

export PATH="$PATH:/Users/williamli/.dotnet/tools"

You can invoke the tool using the following command: dotnetsay
Tool 'dotnetsay' (version '2.1.4') was successfully installed.
```

### CODE REVIEW

**Comment by wli3** [comment]
@KathleenDollard to check the user facing string. Once you copy paste the above long command you need to run `zsh -l` or relogin to make it take effect

File: src/dotnet/ShellShim/OsxBashEnvironmentPath.cs [hunk 1]
```diff
@@ -10,7 +10,7 @@
 
 namespace Microsoft.DotNet.ShellShim
 {
-    internal class OSXEnvironmentPath : IEnvironmentPath
+    internal class OsxBashEnvironmentPath : IEnvironmentPath
```
Comment (peterhuene) [inline src/dotnet/ShellShim/OsxBashEnvironmentPath.cs]: Do we really need two types to implement a slightly different message?  Can't we keep this as-is and change `PrintAddPathInstructioNifPathDoesNotExist` to:

```c#
public void PrintAddPathInstructionIfPathDoesNotExist()
{
    if (!PackageExecutablePathExists())
    {
        if (_fileSystem.Exists(DotnetCliToolsPathsDPath))
        {
            _reporter.WriteLine(
                CommonLocalizableStrings.EnvironmentPathOSXNeedReopen);
        }
        else
        {
            // similar to https://code.visualstudio.com/docs/setup/mac
            _reporter.WriteLine(
                string.Format(
                    new ZshDetector(_environmentProvider).IsZshTheUsersShell() ?
                        CommonLocalizableStrings.EnvironmentPathOSXZshManualInstructions :
                        CommonLocalizableStrings.EnvironmentPathOSXBashManualInstructions,
                    _packageExecutablePath.Path));
        }
    }
}
```

File: src/dotnet/ShellShim/ZshDetector.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        private readonly IEnvironmentProvider _environmentProvider;
+        private const string ZshFileName = "zsh";
+
+        public ZshDetector(IEnvironmentProvider environmentProvider = null)
+        {
+            _environmentProvider = environmentProvider;
+        }
+
+        /// <summary>
+        ///     Return if the user use zsh as "The user's shell" instead of the current shell.
```
Comment (peterhuene) [inline src/dotnet/ShellShim/ZshDetector.cs]:
```suggestion
        /// Returns true if the `SHELL` environment variable ends with `zsh` for the filename.
```

File: src/dotnet/ShellShim/OsxBashEnvironmentPath.cs (hunk 1 above)
Comment (peterhuene) [reply src/dotnet/ShellShim/OsxBashEnvironmentPath.cs]: I assume macOS Catalina will keep paths.d working for zsh, right?

File: src/dotnet/CommonLocalizableStrings.resx [hunk 3]
```diff
@@ -563,6 +563,22 @@ EOF
 
 You can add it to the current session by running the following command:
 
+export PATH="$PATH:{0}"
+</value>
+  </data>
+  <data name="EnvironmentPathOSXZshManualInstructions" xml:space="preserve">
+    <value>Tools directory '{0}' is not currently on the PATH environment variable.
+If you are using bash, you can add it to your profile by running the following command:
```
[the remark this answers is not in the document]
Comment (peterhuene) [reply src/dotnet/CommonLocalizableStrings.resx]: Unfortunately the xlfs weren't regenerated.  The downside of clicking GitHub buttons 😄.

File: src/dotnet/ShellShim/OsxBashEnvironmentPath.cs (hunk 1 above)
Comment (wli3) [reply src/dotnet/ShellShim/OsxBashEnvironmentPath.cs]: There are more difference -- `PackageExecutablePathExists` in bash impl consider `~/.dotnet/tools` be good and skip printing. But it is not the case in zsh. Only full path `/home/myuser/.dotnet/tools` will work on zsh.

Also bash has interface `IEnvironmentPath` which as extra `AddPackageExecutablePathToUserPath` method. That does not exist in zsh (we can only print instruction).

So I think it is easier to reason about it with different impl

Comment (peterhuene) [reply src/dotnet/ShellShim/OsxBashEnvironmentPath.cs]: Ah good point, I forgot that we put `~` on the PATH for bash.

Ok, let's keep it two different implementations.

File: src/dotnet/ShellShim/ZshDetector.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        private const string ZshFileName = "zsh";
+
+        public ZshDetector(IEnvironmentProvider environmentProvider = null)
+        {
+            _environmentProvider = environmentProvider;
+        }
+
+        /// <summary>
+        /// Returns true if the `SHELL` environment variable ends with `zsh` for the filename.
+        ///     By detecting $SHELL's value
```
Comment (peterhuene) [inline src/dotnet/ShellShim/ZshDetector.cs]: This line can now be deleted.  I don't like that suggestions can't be multiline (afaik).

File: src/dotnet/ShellShim/ZshDetector.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        /// </summary>
+        public bool IsZshTheUsersShell()
```
Comment (peterhuene) [inline src/dotnet/ShellShim/ZshDetector.cs]: Perhaps this should just be a static method on a static class that takes `IEnvironmentProvider`?

File: src/dotnet/ShellShim/OsxBashEnvironmentPath.cs (hunk 1 above)
Comment (wli3) [reply src/dotnet/ShellShim/OsxBashEnvironmentPath.cs]: > I assume macOS Catalina will keep paths.d working for zsh, right?

There is no paths.d on macOS only path_helper.
and `path_helper` will not expand `~` like paths.d do. So it end up literal `~/.dotnet/tools` on PATH. which is good for bash but not for zsh

File: src/dotnet/ShellShim/ZshDetector.cs (hunk 5 above)
Comment (wli3) [reply src/dotnet/ShellShim/ZshDetector.cs]: no strong opinion on either. I'll go yours

Comment (wli3) [reply src/dotnet/ShellShim/ZshDetector.cs]: Made it static

File: src/dotnet/ShellShim/ZshDetector.cs (hunk 4 above)
Comment (wli3) [reply src/dotnet/ShellShim/ZshDetector.cs]: deleted the extra line

**Pull request merged**